### PR TITLE
Migrate Coded Sheets to DataBase

### PR DESCRIPTION
**Pull request by ascii-dev**
## Description
Migrate coded data sheets to the app's database

**Note: there are still some little cleaning to do here and there before merging, but those wouldn't stop the code and functionality from being reviewed.**

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## Screenshots
image

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by esirK — COMMENTED** [review]
@ascii-dev I have tried to run the script but nothing has been created in the database. I'm getting the following errors
``` 
exception happened = 'str' object has no attribute 'get'
```
is there something I might be missing?

**Comment by ascii-dev** [comment]
> @ascii-dev I have tried to run the script but nothing has been created in the database. I'm getting the following errors
> 
> ```
> exception happened = 'str' object has no attribute 'get'
> ```
> 
> is there something I might be missing?

Yes @esirK, I noticed this last night and I already started working on it this morning. While working on it though, I uncovered a couple other problems. Already pushing patches to get them resolved but I'll be having a call with Jean in a couple minutes to get her view on what a permanent fix looks like.

**Comment by ascii-dev** [comment]
@esirK the errors you experienced and others discovered as been fixed. Kindly review. Thanks 🙏

**Comment by esirK** [comment]
> @esirK the errors you experienced and others discovered as been fixed. Kindly review. Thanks 🙏
@ascii-dev 
Data is now being created in the database but some fields aren't being populated.
e.g 
Screenshot 2020-11-04 at 05 42 22
Same for other tabs.

**Comment by esirK** [comment]
@ascii-dev  I have created a PR that builds upon what you had done and fixes the issues above including creating Journalists and Multiple people in the news. I created a new PR since the approach I used is a little different from what you have. Kindly review it and let me what you think.
https://github.com/CodeForAfrica/gmmp/pull/110

**Comment by esirK** [comment]
@kilemensi  I think this is ready for your review.

**Review by kilemensi — COMMENTED** [review]
:+1: @ascii-dev  / @esirK 

---------------------------------------

Still trying to wrap my head around our approach.

1. I see we're passing `filename` around a lot... Why is this? I'm under the impression the process would have been like open file, read everything we can, close the file. But looking just at `transfer_data`, we're passing `filename` to `get_journalist`, `get_people`, and `get_sheet`, and in turn, each of these functions open the same file again to read specific info from it.
2. From today's call, it seems like Jean is still working on part of the reading logic. Is it correct to assume that her changes (if any) have direct impact on this code?

**Comment by esirK** [comment]
> 👍 @ascii-dev / @esirK
> 
> Still trying to wrap my head around our approach.
> 
> 1. I see we're passing `filename` around a lot... Why is this? I'm under the impression the process would have been like open file, read everything we can, close the file. But looking just at `transfer_data`, we're passing `filename` to `get_journalist`, `get_people`, and `get_sheet`, and in turn, each of these functions open the same file again to read specific info from it.
> 2. From today's call, it seems like Jean is still working on part of the reading logic. Is it correct to assume that her changes (if any) have direct impact on this code?

@kilemensi 
For the `filename`, I'm not totally sure why Jean went with this approach. 
For the updates that Jean is making, yes it will partially affect this code. From the conversation, it seemed we will only be changing the following lines 
https://github.com/CodeForAfrica/gmmp/pull/109/files#diff-c346c425f1e3773c0ef287bfcd17383cdf27d9d04cad7fe541534bfa0b750a10R36
from using strings like `NewspaperCoding` to `Print`.

**Comment by kilemensi** [comment]
> 
> @kilemensi
> For the `filename`, I'm not totally sure why Jean went with this approach.

I hear you @esirK but Jean is not a technologist.. she's an analyst. So we should take her code as a specification/rough draft of what needs to happen.

If the current GMMP code was written as what I see in this PR, do you think we'd have been able to make modifications/extend it as much as we have done?

>
> For the updates that Jean is making, yes it will partially affect this code. From the conversation, it seemed we will only be changing the following lines
> https://github.com/CodeForAfrica/gmmp/pull/109/files#diff-c346c425f1e3773c0ef287bfcd17383cdf27d9d04cad7fe541534bfa0b750a10R36
> from using strings like `NewspaperCoding` to `Print`.

Cool, :crossed_fingers: the changes will be that small.

**Comment by esirK** [comment]
> > 
> > @kilemensi
> > For the `filename`, I'm not totally sure why Jean went with this approach.
> 
> I hear you @esirK but Jean is not a technologist.. she's an analyst. So we should take her code as a specification/rough draft of what needs to happen.
> 
> If the current GMMP code was written as what I see in this PR, do you think we'd have been able to make modifications/extend it as much as we have done?
> 
> >
> > For the updates that Jean is making, yes it will partially affect this code. From the conversation, it seemed we will only be changing the following lines
> > https://github.com/CodeForAfrica/gmmp/pull/109/files#diff-c346c425f1e3773c0ef287bfcd17383cdf27d9d04cad7fe541534bfa0b750a10R36
> > from using strings like `NewspaperCoding` to `Print`.
> 
> Cool, :crossed_fingers: the changes will be that small.

True @kilemensi .
I haven't spent time trying to improve her script but will do so from the final script that she uploaded to the drive.

File: sheets/utils.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+        "channel": channel,
+        "start_time": start_time,
+        "media_name": media_name,
+        "twitter_handle": twitter_handle,
+        "retweet": retweet,
+        "equality_rights": 'Y' if equality_rights == 'Y' else 'N',
+        "about_women": 'Y' if about_women == 'Y' else 'N',
+        "inequality_women": int(inequality_women) if inequality_women else "",
+        "stereotypes": int(stereotypes) if stereotypes else '',
+        "url_and_multimedia": comments or "N/A", # TODO: Isaiah, I'm not sure why we have comments here, instead of url_and_multimedia
```
Comment (esirK) [inline sheets/utils.py]: @ascii-dev there's one sheet that uses `url_and_multimedia` as comments i.e the `url_and_multimedia`are the comments.

**Review by kilemensi — COMMENTED** [review]
:+1: @esirK / @ascii-dev 

Here are my quick thoughts around our approach.

## High level thoughts

1. App name should probably be **Coding Sheets** rather than just **Sheets**
2. Need to indicate success/failure for each Excel file & coding sheet processed.

## Technical thoughts

### Project

1. Do we still need `djangorestframework` if we're not doing APIs?

### Input
1. Input location should be a _folder_ or _filename_ i.e. if it's a _folder_, load all Excel sheets from it; if it's a _filename_ **and** it's an Excel file, process that file only.
2. If no input, process Excel files from a **know location** which should be configurable i.e. `env` with a `default`

### Data

1. Why do we need special questions to be null _**now**_? How has the dashboard been handling them so far?
2. Include sample/template Excel sheet so that one is able to follow the code while looking at the spreadsheet.

### Extraction/Import

1. `extractor_script.py` should probably be `coding_sheet.py` so that we have methods like `from coding_sheet import get_basic_info`, etc.
2. I think there is a bit too much copying and repeating going on e.g. from management command, we make copies when calling `get_people`, but in `get_people` there is also a call to `add_coding_info` which also makes a copy. This is also repeated when we call other methogs while processing the same sheet, e.g. `get_sheet`.

Once we decide on :point_up: , I think we'll be on our way...

**Comment by esirK** [comment]
@kilemensi  kudos for a good review. Here are some of my thoughts.
> Here are my quick thoughts around our approach.
> 
> ## High level thoughts
> 1. App name should probably be **Coding Sheets** rather than just **Sheets**
> 2. Need to indicate success/failure for each Excel file & coding sheet processed.
1. Yes we can update that.
2. I think one thing we should add is Sentry then we can also be writing the failures/warnings to a file for offline analysis. I would also love to hear your thoughts on how we should handle the failures/warnings
>### Project
> Do we still need djangorestframework if we're not doing APIs?

Right now yes since we are using it's serializer here https://github.com/CodeForAfrica/gmmp/blob/data_transfer_from_sheets/sheets/serializers.py
another option is to update the code to use django serializers

>Input location should be a folder or filename i.e. if it's a folder, load all Excel sheets from it; if it's a filename and it's an Excel file, process that file only.
>If no input, process Excel files from a know location which should be configurable i.e. env with a default

Right now the input location is preconfigured. If someone wants to only pick a single file from the location, they use the `--filename` to specify; Somehow I say this should be good enough instead of allowing people to place the files anywhere in the system.

**Comment by kilemensi** [comment]
> Right now the input location is preconfigured. If someone wants to only pick a single file from the location, they use the `--filename` to specify; Somehow I say this should be good enough instead of allowing people to place the files anywhere in the system.

Bigger picture @esirK , bigger picture.

The Excel files will come from the cloud right (GDrive, OneDrive, etc.)? Usually one uses a client to "sync" their cloud drive with their PC file system. This sync location is again usually somewhere in the home directory i.e. outside GMMP repo location. Our current approach: 
  1. Forces them to manually copy the folders into GMMP repo for the command to work, otherwise they can only specify one file at  at time (or God forbid, manually download folders from cloud drive into GMMP repo location),
  2. Unnecessarily distinguishes between a file and directory: as long we allow user to specify input location, we should accept either since it's pretty much a one liner to check if it's  a file or not. Even 40 year old commands like `cp` in your shell handle both file and folder paths interchangeably.

>  I think one thing we should add is Sentry then we can also be writing the failures/warnings to a file for offline analysis. I would also love to hear your thoughts on how we should handle the failures/warnings

For now, any failure means us notifying the analyst with all the information we can gather from the app and let the analyst either fix the file and try again or manually upload data to the dashboard.

Combining the above two, may be I should also go ahead and ask (I didn't want to complicate things any further): Do we in any shape or form handle re-runs i.e. 
  1. Can we distinguish uploaded Excel files so that we don't process them if the program is run against the same files, or
  2. Can we update the coded stories rather than creating new stories if the program is run against the same set of files?

> Right now yes since we are using it's serializer here https://github.com/CodeForAfrica/gmmp/blob/data_transfer_from_sheets/sheets/serializers.py
> another option is to update the code to use django serializers

If we can, then we should switch to `django` serializers and remove `djangorestframework` as dependency.

**Comment by esirK** [comment]
> > Right now the input location is preconfigured. If someone wants to only pick a single file from the location, they use the `--filename` to specify; Somehow I say this should be good enough instead of allowing people to place the files anywhere in the system.
> 
> Bigger picture @esirK , bigger picture.
> 
> The Excel files will come from the cloud right (GDrive, OneDrive, etc.)? Usually one uses a client to "sync" their cloud drive with their PC file system. This sync location is again usually somewhere in the home directory i.e. outside GMMP repo location. Our current approach:
> 
> 1. Forces them to manually copy the folders into GMMP repo for the command to work, otherwise they can only specify one file at  at time (or God forbid, manually download folders from cloud drive into GMMP repo location),
> 2. Unnecessarily distinguishes between a file and directory: as long we allow user to specify input location, we should accept either since it's pretty much a one liner to check if it's  a file or not. Even 40 year old commands like `cp` in your shell handle both file and folder paths interchangeably.
> 
> > I think one thing we should add is Sentry then we can also be writing the failures/warnings to a file for offline analysis. I would also love to hear your thoughts on how we should handle the failures/warnings
> 
> For now, any failure means us notifying the analyst with all the information we can gather from the app and let the analyst either fix the file and try again or manually upload data to the dashboard.
> 
> Combining the above two, may be I should also go ahead and ask (I didn't want to complicate things any further): Do we in any shape or form handle re-runs i.e.
> 
> 1. Can we distinguish uploaded Excel files so that we don't process them if the program is run against the same files, or
> 2. Can we update the coded stories rather than creating new stories if the program is run against the same set of files?
> 
> > Right now yes since we are using it's serializer here https://github.com/CodeForAfrica/gmmp/blob/data_transfer_from_sheets/sheets/serializers.py
> > another option is to update the code to use django serializers
> 
> If we can, then we should switch to `django` serializers and remove `djangorestframework` as dependency.

@kilemensi this makes perfect sense. 
From this, we have 3 tasks 

1. The files location
2. Identifying uniqueness of a story (Open to ideas on how to do this)
3.  switch to `django` serializer
cc @ascii-dev

**Comment by kilemensi** [comment]
>     2. Identifying uniqueness of a story (Open to ideas on how to do this)
> 
Still thinking about this @esirK  but can we start by checking in the 2015 sheets/code as Sarah suggested... They may have already had a way to uniquely identify a story/sheet?

**Comment by esirK** [comment]
@ascii-dev QQ. Did you manage to test the multi-line comments issue? I tried it with the following sample and It didn't work 
Screenshot 2020-11-27 at 09 10 43

**Comment by ascii-dev** [comment]
> @ascii-dev QQ. Did you manage to test the multi-line comments issue? I tried it with the following sample and It didn't work
> Screenshot 2020-11-27 at 09 10 43

@esirK yes, it works and I've shared the sample sheet I worked with, with you. cc @kilemensi

**Comment by esirK** [comment]
> > Right now the input location is preconfigured. If someone wants to only pick a single file from the location, they use the `--filename` to specify; Somehow I say this should be good enough instead of allowing people to place the files anywhere in the system.
> 
> Bigger picture @esirK , bigger picture.
> 
> The Excel files will come from the cloud right (GDrive, OneDrive, etc.)? Usually one uses a client to "sync" their cloud drive with their PC file system. This sync location is again usually somewhere in the home directory i.e. outside GMMP repo location. Our current approach:
> 
> 1. Forces them to manually copy the folders into GMMP repo for the command to work, otherwise, they can only specify one file at a time (or god forbid, manually download folders from cloud drive into GMMP repo location),
> 2. Unnecessarily distinguishes between a file and directory: as long we allow the user to specify input location, we should accept either since it's pretty much a one-liner to check if it's a file or not. Even 40-year-old commands like `cp` in your shell handle both file and folder paths interchangeably.

@kilemensi  I started implementing this but I got some questions.
1. Who are we targeting to be running the management command?
2. Given that our production is running using docker, if the user specifies the folder containing the excel sheets as `/home/ubuntu/sheets`, this folder will not be available inside the docker container. Will, we thus be required to mount the whole `/home` folder into the container?

**Comment by kilemensi** [comment]
> 
> @kilemensi I started implementing this but I got some quest...

Hey @esirK 

1. Docker & Dokku is our own addition to the GMMP toolchain... the app used to work without them and even now, it should be able to work without them. This means presence/absence of Docker/Dokku shouldn't impact our thinking around features...
2. The target users of GMMP app (at least the command line part of it) is the Tech people such as us or analyst with enough tech background such as Jean.

Combing 1. and 2., whether someone has to mount a whole home folder (I don't know why they'd do that when Docker allows you to mount just a folder you're interested in or they can just create a virtual env and run the app without Docker, just point DB to production DB, if they need to work with multiple folders without having to mount them) boils down to the same requirement: 

> Given a path that the app has read access to, can we process all Excel files from it? If path points to one Excel file, process that one file; if however the path points to a folder, the process all Excel files in that folder.

**Comment by esirK** [comment]
Hello @kilemensi @ascii-dev 
I think this PR is now ready for merging but I would like to hear your thoughts.

File: coding_sheets/apps.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from django.apps import AppConfig
+
+
+class SheetsConfig(AppConfig):
```
Comment (kilemensi) [inline coding_sheets/apps.py:4]:
```suggestion
class CodingSheetsConfig(AppConfig):
```

File: coding_sheets/models.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from django.db import models
+
+
+class ProcessedSheet(models.Model):
+    country = models.CharField(max_length=255)
+    sheet_name = models.CharField(max_length=255)
+    sheet_tab = models.CharField(max_length=255)
+    sheet_row = models.IntegerField()
```
Comment (kilemensi) [inline coding_sheets/models.py:8]: If we're storing even rows, shouldn't this model be called `ProcessedSheetRow` then?

File: coding_sheets/extractor_script.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
+import pandas as pd
+import json
+from django.conf import settings
+import coding_sheets.coding_info as coding_info
+
+pd.set_option("display.max_colwidth", 100)
```
Comment (kilemensi) [inline coding_sheets/extractor_script.py]: What does this do/add?

**Review by kilemensi — COMMENTED** [review]
👍🏾 @esirK 

------------------

My issues with this PR is the code; It's very hard to follow / feels disjoint / etc.

File: coding_sheets/extractor_script.py (hunk 4 above)
Comment (esirK) [reply coding_sheets/extractor_script.py]: This had been added to help in displaying data in the terminal. I've removed it since we no longer need it.

File: coding_sheets/models.py (hunk 3 above)
Comment (esirK) [reply coding_sheets/models.py:8]: sure. Sounds better.
